sql/src/main/java/io/crate/execution/engine/indexing/ShardUpsertRequestAndResponse.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.engine.indexing;
+
+import io.crate.execution.dml.ShardResponse;
+import io.crate.execution.dml.upsert.ShardUpsertRequest;
+
+public class ShardUpsertRequestAndResponse {
+    final ShardUpsertRequest shardRequest;
+    ShardResponse shardResponse;
```

Comment:
Is it really safe to have this being mutable?

As far as I can see it is used in the `ShardResponseActionListener.onResponse` which is called concurrently

Reply:
the `onResponse` is not called concurrently as a dedicated listener is created for each request. anyway I'm may remove that also while refactoring to use a dedicated source-uri-per-item structure.